### DISCUSSION TITLE
Support for Universal Scene Description

### DISCUSSION THREAD

**Discussion by Sythelux** [godotengine/godot-proposals#5469] [Editor]
Universal Scene Description (USD) is now open source. It was developed by pixar, but is open for a while and has increasing popularity.

We should add support for USD files to Godot

https://github.com/PixarAnimationStudios/USD

(I thought we had a discussion about it somewhere, but wasn't able to find it under the keywords of "universal scene description" )

**Comment by Calinou** [comment]
See also https://github.com/godotengine/godot-proposals/issues/714.

USD is a very complex specification, so this will take a while to materialize if ever.

**Comment by RedMser** [comment]
I'd imagine USD could also be a better intermediate format than for example glTF.
In the far future, this would allow Blender files loaded into Godot to carry more information, such as object visibility, n-gons and more light properties. But as it stands, the exporter only creates very basic output.

**Reply by Calinou** [reply]
N-gons are a bad idea in real-time rendering in general, as GPUs can only render triangles directly. You want the geometry within the interchange file to be quads or ideally triangles, so they can be triangulated in a reliable, deterministic way. You don't want automatic triangulation on import to cause a mess on your UV layout or normals :slightly_smiling_face:

**Reply by RedMser** [reply]
Yeah I did mean quads, which is better than the tris support of glTF (not taking extensions into account).
Would help for tesselation and subdivision though, I can imagine.

**Reply by RichardEllicott** [reply]
this should be concidered again, seriously as it appears the industry is looking seriously at OpenUSD. Nvidia have launched a cloud collaboration platform based around it.

just if you could please get them to mention/muse upon it, whether OpenUSD looks like any cop for the future

i feel the gLTF workflow, the blender asset workflow is one of the weakest things we have

gLTF i think is designed as an open asset to be an asset export format that's designed to be lightweight and one way (all quads/ngons destroyed) it severly lacks features the most annoying to me being that all instances are turned into meshes.... so no instances... this means if i make 20 window instances, when it has got to godot it is 20 seperate meshes with no relation to each other or understanding that they are the same.... it is obviously intended to be, and just, an asset export format.... not a scene format

the light are also abysmal, i myself am resigned to the lights just needing to be done in Godot but wow wouldn't it be nice if the lights went through

from what i understand my equivilent Unreal engine user can have a connection live from 3D Max, updating materials, lights everything straight into Unreal on the fly... links to substance designer etc... i'd just like instances to be supported

but i get OpenUSD might look like a sledgehammer to the percieved requirements we have, and also giant and complex... and furthermore if Blender could be taking it more seriously... i feel like if they did we would

but alas i'm just lamanting maybe not protesting it'd be nice to get opinions on this though if it offers anything for us....btw i don't think triangulating a different way will mess up the UV map, the vertexes are the same and UV coordinates are at vertexs, but now you mention it i wonder maybe there is a standard way they traingulate in this format... yeah we'd not be expecting to edit the file futher, we would be loading it in

**Reply by Sythelux** [reply]
I think we need both eventually. I see specific use cases where you use GLTF (runtime asset loading from remote, fast and easy interaction if components) or USD (big fotorealistic rendering scenes and model instances).

**Reply by daniel080400** [reply]
> i feel the gLTF workflow, the blender asset workflow is one of the weakest things we have
> 

I agree that 3d model asset workflow is bad in Godot, but it has nothing to do with glTF or any other exchanging formats

> gLTF i think is designed as an open asset to be an asset export format that's designed to be lightweight and one way (all quads/ngons destroyed) it severly lacks features the most annoying to me being that all instances are turned into meshes.... so no instances... this means if i make 20 window instances, when it has got to godot it is 20 seperate meshes with no relation to each other or understanding that they are the same.... it is obviously intended to be, and just, an asset export format.... not a scene format

You might have the wrong idea here, glTF does store mesh instances, and Godot does treat them as instances. Here's what happening under the hood:

glTF stores your 3d object as "node" and "mesh", for example, if you have 20 windows (Assume you have them share linked to a mesh, ALT+D duplicated them), when saving them into a glTF file, it is stored as 20 nodes linked to 1 single mesh. Then if you import this glTF file into Godot, it recognizes those 20 nodes, creates 20 MeshInstance3D nodes, and links them into a shared mesh resource, there's no duplicated mesh data, just duplicated nodes.

So, the main problem here has nothing to do with glTF, it's how Godot treats glTF when importing. If you mean it should batch 20 windows as a MultiMeshInstance3D, it can be accomplished by writing a import/ export script with ~300 lines of code.

> the light are also abysmal, i myself am resigned to the lights just needing to be done in Godot but wow wouldn't it be nice if the lights went through

Lights have different units and renderer in every software, so exporting light within glTF can be a bad idea, this is why most game engine workflows encourages you to use their built-in lighting. 

> from what i understand my equivilent Unreal engine user can have a connection live from 3D Max, updating materials, lights everything straight into Unreal on the fly... links to substance designer etc... i'd just like instances to be supported

That's what Godot trying to do with Blender. As you can see, the better approach is to just use .blend or .3dmax files directly if all you want is exporting a full scene to game engines, not seperate meshes.
